### PR TITLE
feat: Generate deduplicate merge object [DHIS2-11562]

### PR DESCRIPTION
**Pull request by abyot**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java [hunk 1]
```diff
@@ -59,7 +60,40 @@
...
+            .map( teav -> teav.getAttribute().getUid() )
+            .filter( tea -> !existingTrackedEntityAttributes.contains( tea ) )
+            .collect( Collectors.toList() );
+
+        if ( !trackedEntityAttributes.isEmpty() )
+        {
+            mergeObject.setTrackedEntityAttributes( trackedEntityAttributes );
+        }
+
+        Set<String> existingRelationships = original.getRelationshipItems().stream()
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:83]: original.getRelationshipItems() would return a list of relationshipItems that reference original, so all the uids you are getting is the same, namely original's uid.

Instead, if we are getting all relationships that are in duplicate but not original, you would do
original.getRelationshipItems().stream().map( ri -> ri.getRelationship().getUid() ). This way we get the uid of the relationship itself, which we can compare.

If the relationship uid is the same, that implies both original and duplicate are part of the relationship, so we skip it :)

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java [hunk 2]
```diff
@@ -59,7 +60,40 @@
...
+            .map( rel -> rel.getTrackedEntityInstance().getUid() ).collect( Collectors.toSet() );
+
+        List<String> relationships = duplicate.getRelationshipItems().stream()
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:86]: See previous comment about getting Relationship uid, not tei uid :)

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java [hunk 3]
```diff
@@ -59,7 +60,40 @@
 
     public MergeObject generateMergeObject( TrackedEntityInstance original, TrackedEntityInstance duplicate )
     {
-        return null;
+        if ( duplicate == null )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:63]: We can assume original and duplicate are never null, due to the flow; We check if both exist in the controller. However, assume we dont know that, I would make sure both are not null :)

**Review by enricocolasante — COMMENTED** [review]
We have a class DeduplicationHelperTest where we can unit test this method.
It would really help to understand the logic and to be sure it will not be broken by future changes on the code

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java [hunk 4]
```diff
@@ -59,7 +64,28 @@
...
+
+        MergeObject mergeObject = MergeObject.builder()
+            .trackedEntityAttributes( attributes )
+            .relationships( relationships )
+            .build();
+
+        if ( mergeObject.getTrackedEntityAttributes().isEmpty() && mergeObject.getRelationships().isEmpty() )
+        {
+            throw new PotentialDuplicateNullException(
+                "Potentical Duplicate is null hence nothing to merge" );
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:85]: Does it make sense to throw an error if there is nothing to merge?
Lets say we have to identical TEIs, we still want the duplicate to be removed and the potentialDuplicate to be marked as MERGED

**Review by enricocolasante — CHANGES_REQUESTED** [review]
I am not sure about the logic around PotentialDuplicateNullException

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java (hunk 4 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:117]: not sure. what shall we do when the item to be merged has nothing to be merged?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java (hunk 3 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java]: check for null is now removed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java (hunk 4 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:117]: If the mergeObject is empty, it implies we keep the original and delete the duplicate :) No error needed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java [hunk 5]
```diff
@@ -138,4 +164,70 @@ public boolean hasUserAccess( TrackedEntityInstance original, TrackedEntityInsta
...
+    private List<String> getMergeableRelationships( TrackedEntityInstance original, TrackedEntityInstance duplicate )
+    {
+        Map<String, Relationship> existingRelationships = original.getRelationshipItems().stream()
+            .collect( Collectors.toMap( ri -> ri.getRelationship().getUid(), RelationshipItem::getRelationship ) );
+
+        List<String> relationships = new ArrayList<>();
+
+        for ( RelationshipItem ri : duplicate.getRelationshipItems() )
+        {
+            Relationship existing = existingRelationships.get( ri.getRelationship().getUid() );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java:205]: The relationship uid is unique for all relationships, so what you are doing in your first if ( existing != null ) is saying if both original and duplicate have the same relationship. Which is great, and all you need. If they share uid, its means its a relationship between original and duplicate, which is the only one we block for. That also means you dont need the second equals check(ri.getRelationship) and you dont need the else block (TrackedEntityInstance from = ..)

This was a much better way to do what I explained complicated yesteday 😅

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java (hunk 4 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java]: empty mergeObject is now returned

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/PotentialDuplicateNullException.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+/**
+ * @author Abyot Asalefew Gizaw <abyota@gmail.com>
+ *
+ */
+@Getter
+@ResponseStatus( HttpStatus.FORBIDDEN )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/PotentialDuplicateNullException.java:40]: We don't need this exception anymore

**Review by enricocolasante — APPROVED** [review]
Remove PotentialDuplicateNullException.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/PotentialDuplicateNullException.java (hunk 6 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/PotentialDuplicateNullException.java:40]: yes, true.

**Pull request merged**